client/lcd/root.go

```diff
@@ -1,36 +1,157 @@
 package lcd
 
 import (
-	"errors"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"log"
+	"math/big"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
 
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
 
 	"github.com/cosmos/cosmos-sdk/client"
+	wire "github.com/cosmos/cosmos-sdk/wire"
 )
 
 const (
-	flagBind = "bind"
-	flagCORS = "cors"
+	flagBind             = "bind"
+	flagCORS             = "cors"
+	flagUnsafeConnection = "unsafe_connection"
 )
 
-// XXX: remove this when not needed
-func todoNotImplemented(_ *cobra.Command, _ []string) error {
-	return errors.New("TODO: Command not yet implemented")
-}
-
 // ServeCommand will generate a long-running rest server
 // (aka Light Client Daemon) that exposes functionality similar
 // to the cli, but over rest
 func ServeCommand() *cobra.Command {
+	// TODO get code from app
+	cdc := wire.NewCodec()
 	cmd := &cobra.Command{
-		Use:   "serve",
+		Use:   "rest-server",
 		Short: "Start LCD (light-client daemon), a local REST server",
-		RunE:  todoNotImplemented,
+		RunE:  startRESTServer(cdc),
 	}
 	// TODO: handle unix sockets also?
 	cmd.Flags().StringP(flagBind, "b", "localhost:1317", "Interface and port that server binds to")
 	cmd.Flags().String(flagCORS, "", "Set to domains that can make CORS requests (* for all)")
+	cmd.Flags().String(flagUnsafeConnection, "false", "Do not enable HTTPS for the REST server")
 	cmd.Flags().StringP(client.FlagChainID, "c", "", "ID of chain we connect to")
 	cmd.Flags().StringP(client.FlagNode, "n", "tcp://localhost:46657", "Node to connect to")
 	return cmd
 }
+
+func startRESTServer(cdc *wire.Codec) func(cmd *cobra.Command, args []string) error {
+	return func(cmd *cobra.Command, args []string) error {
+		bind := viper.GetString(flagBind)
+		unsafeConnection, err := strconv.ParseBool(viper.GetString(flagUnsafeConnection))
+		if err != nil {
+			return err
+		}
+		r := initRouter(cdc)
+
+		if unsafeConnection {
+			return http.ListenAndServe(bind, r)
+		}
+
+		// setup https
+		// get path to certificates
+		ex, err := os.Executable()
+		if err != nil {
+			return err
+		}
+		exPath := filepath.Dir(ex)
+
+		_, err = os.Stat(filepath.Join(exPath, "server.crt"))
+		os.IsNotExist(err)
+		_, err = os.Stat(filepath.Join(exPath, "server.key"))
+		os.IsNotExist(err)
+
+		if err != nil {
+			err = generateAndSaveCertificate(exPath)
```

Comment:
do you use https://letsencrypt.org/ ? or is this just generates raw/unconfirmed certificate?

Reply:
These should not be certificates signed by a public certificate authority. They need to be generated locally.